### PR TITLE
Deployment preparation: `V3 Routers`

### PR DESCRIPTION
**Pull request by jubeira**
# Deployment preparation: `V3 Routers`

Build-info: https://github.com/balancer/balancer-v3-monorepo/commit/74d7068fb21565741427cdabfa4f1b539a4bddaa

All tasks have the same artifact, but I believe it's better to create separate tasks in case we ever deprecate any of them individually.

## Deployment preparation tasks checklist:

- [x] Deployments `README` is updated with the new task name and link
- [x] Task `README` has a description and links to the artifacts
- [x] `build-info` is updated with artifacts compiled at the specified commit
- [x] Artifacts are generated from `build-info`
- [x] New task has a working fork test 

## Code checklist:

- [ ] The diff is legible and has no extraneous changes 
- [ ] Complex code has been commented in tests and scripts
- [ ] Fork test block number and test cases are appropriate for this deployment
- [ ] The base branch is either `master`, or there's a description of how to merge

### CODE REVIEW

**Comment by EndymionJkb** [comment]
Starting to look. I confirm the monorepo commit is correct, and the build-info files match as well.

The build-info in composite-liquidity-router is called `Router`; should be `CompositeLiquidityRouter` to match the others.

File: v3/tasks/20241205-v3-buffer-router/readme.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# V3 Buffer Router
+
+Buffer Router set deployment for V3.
+Contains `BufferRouter` to initialize and manage liquidity positions for buffers.
+
+## Useful Files
+
+- [Code](https://github.com/balancer/balancer-v3-monorepo/commit/74d7068fb21565741427cdabfa4f1b539a4bddaa).
+- [Sepolia testnet addresses](./output/sepolia.json)
```
Comment (EndymionJkb) [inline v3/tasks/20241205-v3-buffer-router/readme.md:9]:
```suggestion
```

File: v3/tasks/20241205-v3-composite-liquidity-router/readme.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# V3 Composite Liquidity Router
+
+Composite Liquidity Router deployment for V3.
+Contains `CompositeLiquidityRouter` for complex liquidity operations involving pools with ERC4626 wrappers and nested pools.
+
+## Useful Files
+
+- [Code](https://github.com/balancer/balancer-v3-monorepo/commit/74d7068fb21565741427cdabfa4f1b539a4bddaa).
+- [Sepolia testnet addresses](./output/sepolia.json)
```
Comment (EndymionJkb) [inline v3/tasks/20241205-v3-composite-liquidity-router/readme.md:9]:
```suggestion
```

File: v3/tasks/20241205-v3-composite-liquidity-router/test/task.fork.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import hre, { ethers } from 'hardhat';
+import { expect } from 'chai';
+import { describeForkTest, getForkedNetwork, getSigner, impersonate, Task, TaskMode } from '@src';
+import { Contract } from 'ethers';
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';
+import { fp } from '@helpers/numbers';
+
+describeForkTest('Routers-V3', 'mainnet', 21336200, function () {
```
Comment (EndymionJkb) [inline v3/tasks/20241205-v3-composite-liquidity-router/test/task.fork.ts:8]:
```suggestion
describeForkTest('CompositeLiquidityRouter-V3', 'mainnet', 21336200, function () {
```

File: v3/tasks/20241205-v3-router/readme.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# V3 Router
+
+Base Router deployment for V3.
+Contains `Router` for basic, single step operation (pool initialization, add, remove, swap).
```
Comment (EndymionJkb) [inline v3/tasks/20241205-v3-router/readme.md:4]:
```suggestion
Contains `Router` for basic, single step operations (e.g., pool initialization, add, remove, swap).
```

File: v3/tasks/20241205-v3-router/readme.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+## Useful Files
+
+- [Code](https://github.com/balancer/balancer-v3-monorepo/commit/74d7068fb21565741427cdabfa4f1b539a4bddaa).
+- [Sepolia testnet addresses](./output/sepolia.json)
```
Comment (EndymionJkb) [inline v3/tasks/20241205-v3-router/readme.md:9]:
```suggestion
```

**Review by EndymionJkb — COMMENTED** [review]
First pass done. Maybe we can port some fork tests over from v2 here? (Or adapt the v3 tests?)

**Comment by jamiegreen7** [comment]
Bro stop it

**Comment by jamiegreen7** [comment]
Bro stop it

**Comment by jamiegreen7** [comment]
I need my money

File: v3/tasks/00000000-permit2/readme.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Permit2
```
[the remark this answers is not in the document]
Comment (jubeira) [reply v3/tasks/00000000-permit2/readme.md:1]: We don't have a fork test there though.

[the remark this answers is not in the document]
Comment (jubeira) [reply v3/tasks/00000000-permit2/readme.md:1]: ah but we'll need it for deployment.

**Review by joaobrunoah — DISMISSED** [review]
I can attest that the build-info of all routers matches the commit 74d7068fb21565741427cdabfa4f1b539a4bddaa of the V3 monorepo.

**Review by joaobrunoah — APPROVED** [review]

File: v3/tasks/20241205-v3-batch-router/readme.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# V3 Batch Router
```
Comment (EndymionJkb) [inline v3/tasks/20241205-v3-batch-router/readme.md:1]:
```suggestion
# 2024-12-05 - V3 Batch Router
```
Didn't notice this on the pools (should do a follow-up for consistency), but we have always included the date in the readme. Here and elsewhere.

**Review by EndymionJkb — COMMENTED** [review]
Except for a readme comment, looks like everything is "aligned" (i.e., similarly named things aren't mixed up).

**Review by joaobrunoah — DISMISSED** [review]
LGTM! I already confirmed that build-info matches the correct commit of the monorepo.

File: v3/tasks/20241205-v3-buffer-router/test/task.fork.ts [hunk 8]
```diff
@@ -47,4 +61,21 @@ describeForkTest('BufferRouter-V3', 'mainnet', 21336200, function () {
...
+  it('tests buffer initialization', async () => {
+    const initBalance = 1000000e6;
+    await usdc.connect(usdcWhale).approve(permit2.address, initBalance);
+    await permit2.connect(usdcWhale).approve(USDC_ADDRESS, bufferRouter.address, initBalance, MAX_UINT48);
+
+    // We need to call methods that don't exist in the vault.
+    const abi = [
+      'function isERC4626BufferInitialized(address) external view returns (bool)',
+      'function getERC4626BufferAsset(address) external view returns (address)',
+    ];
```
Comment (joaobrunoah) [inline v3/tasks/20241205-v3-buffer-router/test/task.fork.ts:74]: Cool!!

**Review by joaobrunoah — APPROVED** [review]

**Review by joaobrunoah — APPROVED** [review]

**Pull request merged**